python_modules/dagster/dagster/core/definitions/assets_info.py

```diff
@@ -0,0 +1,184 @@
+from typing import (
+    TYPE_CHECKING,
+    AbstractSet,
+    Callable,
+    Dict,
+    Mapping,
+    NamedTuple,
+    Optional,
+    Set,
+    Tuple,
+)
+
+from dagster import check
+from dagster.core.definitions.events import AssetKey
+
+from .dependency import NodeHandle, NodeInputHandle, NodeOutputHandle
+from .graph_definition import GraphDefinition
+from .node_definition import NodeDefinition
+from .partition import PartitionsDefinition
+
+if TYPE_CHECKING:
+    from dagster.core.execution.context.output import OutputContext
+
+
+class AssetInfo(
+    NamedTuple(
+        "_AssetInfo",
+        [
+            ("asset_key", AssetKey),
+            ("asset_partitions_fn", Callable[["OutputContext"], Optional[AbstractSet[str]]]),
+            ("asset_partitions_def", Optional[PartitionsDefinition]),
+        ],
+    )
+):
+    """Defines all of the asset-related information for a given input or output.
+
+    Args:
+        asset_key (AssetKey): The AssetKey
+        asset_partitions_fn (OutputContext -> Optional[Set[str]], optional): A function which takes the
+            current OutputContext and generates a set of partition names that will be materialized
+            for this asset.
+        asset_partitions_def (PartitionsDefinition, optional): Defines the set of valid partitions
+            for this asset.
+    """
+
+    def __new__(
```

Comment:
FWIW, if this is only internal, we could get away without the `__new__` and checks because mypy will handle it for us.